src/middlewared/middlewared/plugins/pool.py

```diff
@@ -2749,13 +2813,564 @@ def query(self, filters=None, options=None):
+                'key': key, 'load_key': False,
+            }
+        )
+
+        # TODO: Handle renames of datasets appropriately wrt encryption roots and db - this will be done when
+        #  devd changes are in from the OS end
+        await self.insert_or_update_encrypted_record(
+            {'encryption_key': key, 'key_format': options['key_format'], 'name': id}
+        )
+
+    @accepts(Str('id'))
+    async def inherit_parent_encryption_properties(self, id):
+        """
+        Allows to inherit parent's encryption root discarding it's current encryption settings. It should be noted
+        this can only be done where `id` has an encrypted parent and `id` itself is an encryption root.
+        """
+        ds = await self._get_instance(id)
+        if not ds['encrypted']:
+            raise CallError(f'Dataset {id} is not encrypted')
+        elif not ds['encryption_root'] != id:
+            raise CallError(f'Dataset {id} is not an encryption root')
+        elif ds['locked']:
+            raise CallError('Dataset must be unlocked to perform this operation')
+        elif '/' not in id:
+            raise CallError('This operation is not valid for root datasets')
+        else:
+            parent = (await self._get_instance(id.rsplit('/', 1)[0]))
+            if not parent['encrypted']:
+                raise CallError('This operation requires parent dataset to be encrypted')
+            else:
+                parent_encrypted_root = (await self._get_instance(parent['encryption_root']))
+                if ZFSKeyFormat(parent_encrypted_root['key_format']['value']) == ZFSKeyFormat.PASSPHRASE.value:
+                    if any(
+                        d['name'] == d['encryption_root']
+                        for d in await self.middleware.run_in_thread(
+                            self.query, [
+                                ['id', '^', f'{id}/'], ['encrypted', '=', True],
+                                ['key_format.value', '!=', ZFSKeyFormat.PASSPHRASE.value]
+                            ]
+                        )
+                    ):
+                        raise CallError(
+                            f'{id} has children which are encrypted with HEX/RAW keys. It is not allowed to '
+                            'have HEX/RAW encrypted roots as children when parent encrypted root is encrypted'
+                            'with PASSPHRASE key format.'
+                        )
+
+        await self.middleware.call('zfs.dataset.change_encryption_root', id, {'load_key': False})
+        await self.middleware.call(
+            'pool.dataset.sync_db_keys', ['OR', [['name', '=', id], ['name', '^', f'{id}/']]]
+        )
+
+    @private
+    def _retrieve_keys_from_compressed_file(self, job, ds_name):
+        job.check_pipe('input')
+        temp_dir = key_file = None
+        data = {}
+        try:
+            with tempfile.NamedTemporaryFile(delete=False, mode='wb') as f:
```

Comment:
Why `delete=False` and everything else outside of context manager block?

Reply:
The idea is that if for any reason an exception occurs while we are copying from the job pipe to the temporary fie, we handle that separately and reading from the newly written temp file would also mean that i have to close the previous handle, so we accomplish that as well. An exception in the latter section means that we have been provided with a single key file instead of a tar file.

Reply:
Your `key_file` value is overwritten in line 3360 so `f` is never deleted in line 3368, is this intentional and what for?